File: Python/64-minPathSum.py
class Solution:
    # @param {integer[][]} grid
    # @return {integer}
    def minPathSum(self, grid):
        m = len(grid)
        n = len(grid[0])
        for i in xrange(1, n):
            grid[0][i] += grid[0][i - 1] 
        for i in xrange(1, m):
            grid[i][0] += grid[i - 1][0]
        for i in xrange(1, m):
            for j in xrange(1, n):
                grid[i][j] += min(grid[i - 1][j], grid[i][j - 1])
        # s.print_matrix(grid)
        return grid[-1][-1]

    def print_matrix(self, matrix):
        for i in xrange(len(matrix)):
            print ' '.join([str(n) for n in matrix[i]])

if __name__ == '__main__':
   s = Solution()
   matrix = [             \
                [1, 3, 5],\
                [1, 2, 2],\
                [4, 3, 1],\
            ]
   print s.minPathSum(matrix)
   matrix = [[1, 2, 3, 4, 5]]
   print s.minPathSum(matrix)
   matrix = [[1]]
   print s.minPathSum(matrix)
   matrix = [[1], [1], [1]]
   print s.minPathSum(matrix)

File: C++/64-MinPathSum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int> >& grid) {
        int rowNum = grid.size(), colNum = grid[0].size();
        for (int i = 1; i < rowNum; i++) {
             grid[i][0] += grid[i - 1][0];
        };
        for (int j = 1; j < colNum; j++) {
             grid[0][j] += grid[0][j - 1];
        };
        for (int i = 1; i < rowNum; i++) {
             for (int j = 1; j < colNum; j++) {
                  grid[i][j] += grid[i - 1][j] > grid[i][j - 1] ? grid[i][j - 1] : grid[i - 1][j];
             };
        };
        return grid[rowNum - 1][colNum - 1];
    }
};